File: Database/API/search-foods.php
<?php

class IndexLenPair {
	public $index;
	public $relevance;
	public $popularity;
	function __construct($i, $l, $p) {
		$this->index = (int) $i;
		$this->relevance = (int) $l;
		$this->popularity = (int) $p;
	}
}

function getFoodData($index) {
	$path = "../Database-Files/Food-Data/" . $index . ".txt";
	if (file_exists($path)) {
		$f = fopen($path, "r");
		$file_str = fread($f, filesize($path));
		fclose($f);
		return $file_str;
	}
	else
		return NULL;
}

function getFoodsWithTrichar($trichar) {
	$path = "../Database-Files/Trichars-To-Index/" . $trichar . ".txt";
	if (file_exists($path)) {
		$f = fopen($path, "r");
		$file_str = fread($f, filesize($path));
		$matches = explode(";", $file_str);
		for ($k = 0; $k < count($matches); ++$k) {
			$temp = explode(",", $matches[$k]);
			if (sizeof($temp) == 2)
				$matches[$k] = new IndexLenPair($temp[0], $temp[1], 0);
			else
				$matches[$k] = new IndexLenPair($temp[0], $temp[1], $temp[2]);
		}
		fclose($f);
		return $matches;
	}
	else
		return NULL;
}


$search_str = $_GET["str"];
$max_results_to_return = 50;

$search_str = preg_replace("/[^A-Za-z0-9 ]/", '', $search_str);
$search_arr = split(" ", strtolower($search_str));

$trichars = array();
for ($i = 0; $i < count($search_arr); ++$i) {
	for ($j = 0; $j < strlen($search_arr[$i])-2; ++$j) {
		$current_trichar = substr($search_arr[$i], $j, 3);
		$trichars[$current_trichar] = 1;
	}
}

$results = array();
$popularity = array();

foreach ($trichars as $trichar => $value) {
	$trichar_matches = getFoodsWithTrichar($trichar);
	if ($trichar_matches == NULL)
		continue;
	foreach ($trichar_matches as &$pair) {
		$pair->index = $pair->index . "";
		if (array_key_exists($pair->index, $results))
			$results[$pair->index] = $results[$pair->index] + $pair->relevance;
		else {
			$results[$pair->index] = $pair->relevance;
			$popularity[$pair->index] = $pair->popularity;
		}
	}
}

$results_list = [];
foreach ($results as $index => $val)
	array_push($results_list, [(int) $index, $results[$index] * ($popularity[$index] + 1)]);

function foo($bar, $baz) {
	return $bar[1] < $baz[1] ? 1 : -1;
}

usort($results_list, "foo");

$output = "[";
for ($i = 0; $i < $max_results_to_return; ++$i) {
	if ($i >= count($results_list))
		break;
	if ($i != 0)
		$output .= ",";
	$food_info = getFoodData($results_list[$i][0]);
	$output .= "{\"index\": " . $results_list[$i][0] . ", \"info\": " . $food_info . ", \"popularity\": " . $popularity[$results_list[$i][0]] . "}";
}
$output .= "]";

echo $output;
?>

File: Database/API/search-foods.py
import os
import sys
import re

# python search-foods.py <search string> <maximum number of results>

class IndexLenPair:
	def __init__(self, i, l):
		self.index = int(i)
		self.len = int(l)


def getFoodsWithTrichar(trichar):
	path = "../Database-Files/Trichars-To-Index/" + trichar + ".txt"
	if os.path.isfile(path):
		f = open(path, "r")
		matches = f.read().split(";")
		for i in range(0, len(matches)):
			temp = matches[i].split(",")
			matches[i] = IndexLenPair(temp[0], temp[1])
		f.close()
		return set(matches)
	else:
		return None


search_str = ""
max_results_to_return = 20
if len(sys.argv) == 3:
	search_str = sys.argv[1]
	try:
		max_results_to_return = int(sys.argv[2])
	except ValueError:
		print "Error: second argument must be an integer"
		sys.exit(0)
elif len(sys.argv) == 2:
	search_str = sys.argv[1]
elif len(sys.argv) == 1:
	print "Error: No search string given"
	sys.exit(0)
else:
	print "Error: Too many arguments"
	sys.exit(0)

search_str = re.sub('\W+', '', search_str)
search_arr = search_str.lower().split(" ")

trichars = {}
for i in range(0, len(search_arr)):
	for j in range(0, len(search_arr[i])-2):
		current_trichar = search_arr[i][j:j+3]
		trichars[current_trichar] = 1

results = {}

for trichar in trichars:
	trichar_matches = getFoodsWithTrichar(trichar)
	if trichar_matches == None:
		continue
	for pair in trichar_matches:
		if pair.index in results:
			results[pair.index] = results[pair.index] + 1.0/pair.len
		else:
			results[pair.index] = 1.0/pair.len
			

results_list = []
for index in results:
	results_list.append([int(index), results[index]])


def foo(bar):
	return -1 * bar[1]

results_list = sorted(results_list, key = foo)

output = ""
for i in range(0, max_results_to_return):
	if i >= len(results_list):
		break
	if i != 0:
		output += ","
	output += str(results_list[i][0])
print output
